src/passes/MemoryPacking.cpp

```diff
@@ -23,141 +23,394 @@
+        size_t bytesWritten = 0;
+
+        // Split curr into multiple memory.inits and memory.fills, storing the
+        // original base destination in a local if it is not a constant
+        Expression* result = nullptr;
+        Index destVar = -1;
+        if (!curr->dest->is<Const>()) {
+          destVar = builder.addVar(getFunction(), i32);
+          result = builder.makeLocalSet(destVar, curr->dest);
+        }
+        for (auto& range : ranges) {
+          // Skip ranges that do no intersect the range of the memory.fill
+          if (range.end <= start) {
+            if (!range.isZero) {
+              segmentIndex++;
+            }
+            continue;
+          }
+          if (range.start >= end) {
+            break;
+          }
+
+          Expression* dest;
+          if (auto* c = curr->dest->dynCast<Const>()) {
+            dest = builder.makeConst(
+              Literal(int32_t(c->value.geti32() + bytesWritten)));
+          } else {
+            dest = builder.makeLocalGet(destVar, i32);
+            if (bytesWritten > 0) {
+              Const* addend = builder.makeConst(Literal(int32_t(bytesWritten)));
+              dest = builder.makeBinary(AddInt32, dest, addend);
+            }
+          }
+
+          size_t bytes =
+            std::min(range.end, end) - std::max(range.start, start);
+          bytesWritten += bytes;
+          Expression* size = builder.makeConst(Literal(int32_t(bytes)));
+
+          if (range.isZero) {
+            Expression* value = builder.makeConst(Literal::makeZero(i32));
+            result = builder.blockify(
+              result, builder.makeMemoryFill(dest, value, size));
+          } else {
+            size_t offsetBytes = std::max(start, range.start) - range.start;
+            Expression* offset =
+              builder.makeConst(Literal(int32_t(offsetBytes)));
+            result = builder.blockify(
+              result, builder.makeMemoryInit(segmentIndex, dest, offset, size));
+            segmentIndex++;
+          }
+        }
+        if (result) {
+          replaceCurrent(result);
+        } else {
+          // Zero-width fill with constant dest does not intersect any ranges
+          ExpressionManipulator::nop(curr);
+        }
+      }
+
+      void visitDataDrop(DataDrop* curr) {
+        auto replacement = replacements.find(curr);
+        assert(replacement != replacements.end());
+
+        auto segmentIndex = replacement->second.first;
+        auto& ranges = replacement->second.second;
+
+        Builder builder(*getModule());
+
+        Expression* result = nullptr;
+        for (auto range : ranges) {
+          if (!range.isZero) {
+            result =
+              builder.blockify(result, builder.makeDataDrop(segmentIndex++));
+          }
+        }
+
+        if (result) {
+          replaceCurrent(result);
+        } else {
+          // All-zero or empty segment, just update index
+          curr->segment = segmentIndex;
```

Comment:
Are we not emulating trapping behavior here?

Reply:
In this case there was nothing to emulate because the corresponding index was not split.